packages/@aws-cdk/custom-resource-handlers/lib/custom-resources-framework/framework.ts

```diff
@@ -0,0 +1,121 @@
+/* eslint-disable import/no-extraneous-dependencies */
+import { ExternalModule, InterfaceType, Module, TypeScriptRenderer } from '@cdklabs/typewriter';
+import * as fs from 'fs-extra';
+import { CdkCustomResourceClass, CdkCustomResourceClassProps } from './classes';
+import { ProviderType, ProviderProps } from './config';
+import { Runtime } from './runtime';
+import { RuntimeDeterminer } from './utils/runtime-determiner';
+
+export class CdkCustomResourceModule extends Module {
+  /**
+   * The latest nodejs runtime version available across all AWS regions.
+   */
+  private static readonly DEFAULT_RUNTIME = Runtime.NODEJS_18_X;
+
+  private readonly renderer = new TypeScriptRenderer();
+  private readonly externalModules = new Map<string, boolean>();
+  private readonly _interfaces = new Map<string, InterfaceType>();
+  private _hasProviders = false;
+
+  /**
+   * Whether the module being generated will live inside of aws-cdk-lib/core.
+   */
+  public readonly coreInternal: boolean;
+
+  /**
+   * Whether the module contains custom resource providers.
+   */
+  public get hasProviders() {
+    return this._hasProviders;
+  }
+
+  public constructor(fqn: string) {
+    super(fqn);
+    this.coreInternal = fqn.includes('core');
+  }
+
+  /**
+   * Build a framework component inside of this module.
+   */
+  public build(provider: ProviderProps, codeDirectory: string) {
+    if (provider.type === ProviderType.CDK_NO_OP) {
+      return;
+    }
+
+    this._hasProviders = true;
+
+    const handler = provider.handler ?? 'index.handler';
+    const name = this.buildProviderName(handler);
+
+    const props: CdkCustomResourceClassProps = {
+      name,
+      handler,
+      codeDirectory,
+      runtime: RuntimeDeterminer.determineLatestRuntime(
+        provider.compatibleRuntimes,
+        CdkCustomResourceModule.DEFAULT_RUNTIME,
+      ),
```

Comment:
I would really like this code to be handled at runtime, not at codegen time. Because the next step (out of scope for this PR) would be to add Node20 as a compatible runtime everywhere, but keep the "preferred" runtime on Node18 (since CN and GovCloud don't have Node20 yet). Then we can change the `RuntimeDeterminer` to take a custom fact into account, in case someone wants to run them with node20 already.

I'm concerned that if we don't this check at runtime now, we back ourselves into a corner we cannot get out anymore. I might be wrong though. Thoughts?